### ISSUE TITLE
FreeBSD instructions on the web site do not work (cloudabi-run)

### ISSUE THREAD

**Issue by jeking3** [NuxiNL/cloudabi#4]
https://nuxi.nl/cloudabi/freebsd/

'''
[user@freebsd ~]$ sudo pkg install cloudabi-utils
Password:
Updating FreeBSD repository catalogue...
FreeBSD repository is up to date.
Updating CloudABI repository catalogue...
CloudABI repository is up to date.
All repositories are up to date.
pkg: No packages available to install matching 'cloudabi-utils' have been found in the repositories
'''

**Comment by EdSchouten** [comment]
Hey! That's annoying. The `cloudabi-utils` package should be part of the official FreeBSD repository.

What version of FreeBSD are you using and what does your `pkg(8)` config look like?

By the way, as a workaround you can install `cloudabi-utils` from the FreeBSD ports tree.

**Comment by jeking3** [comment]
I downloaded the latest one available as of yesterday.
```
#define __FreeBSD_version 1101001       /* Master, propagated to newvers */
```

For what it's worth, I refactored the random provider inside Boost.Uuid so that it detects and uses the CloudABI system call `cloudabi_sys_random_get` for entropy and passed tests.  Expecting to merge that into Boost 1.67.0 (not the upcoming 1.66.0) because it's a fairly large change under the covers.

I didn't seem to need cloudabi-run in order to run the executables I cross-compiled.

**Comment by EdSchouten** [comment]
> For what it's worth, I refactored the random provider inside Boost.Uuid so that it detects and uses the CloudABI system call cloudabi_sys_random_get for entropy and passed tests.

Nice! Quick question: why did you decide to use `cloudabi_sys_random_get()`? libc's `arc4random_buf()` has the advantage of not calling into the kernel so often. I'm really happy to see someone has stepped up to help me get rid of some of my local changes to Boost. Thanks!

> I didn't seem to need cloudabi-run in order to run the executables I cross-compiled.

That's true. You can cross compile and `./foo` an executable without using `cloudabi-run`, but if you do that, it's harder to pass in file descriptors to resources properly.

You only need to use `cloudabi-run` in case you use `void program_main(const argdata_t *ad)` as the entry point, instead of `int main()`. The `cloudabi-run` man page has some examples of that.

FreeBSD 11... Weird. A binary package should have been available there. I'll look into it.

**Comment by jeking3** [comment]
I wasn't aware that arc4random was accessible.  I already have a code path for arc4random for OpenBSD so I will have the cloudabi stuff use it instead.  I'm the (relatively) new maintainer of Boost.Uuid and Boost.Format and there was a feature request for this either in Boost.Trac or on github.

**Comment by jeking3** [comment]
Well, that was easy to change. :)  Thanks for the tip.

**Comment by EdSchouten** [comment]
You're welcome!

I think I've figured out why `cloudabi-utils` isn't available on your system. I recently had to change this port to depend on more recent versions of 11.x. You now need a version of 11.x somewhere between 11.0 and 11.1. Because [11.0 is still supported until the end of this month](https://www.freebsd.org/security/), binary packages for all of 11.x are still built on 11.0. After 11.0 becomes unsupported, binary packages will be built on 11.1 systems.

Long story short, this is thus an issue that will get resolved automatically. That said, I will extend the installation documentation to mention this.

**Comment by EdSchouten** [comment]
The website has been updated. Thanks again for reporting!

**Comment by jeking3** [comment]
Any changes you have made to modules locally for boost should be conditionalized using preprocessor macros from Boost.Predef:

```
#include <boost/predef/os/cloudabi.h>
#include <boost/predef/lib/c/cloudabi.h>

#if BOOST_OS_CLOUDABI
  ...
#endif

#if BOOST_LIB_C_CLOUDABI >= BOOST_VERSION_NUMBER(0, 68, 0)
  ...
#endif
```

Changes to each repository can be submitted as pull requests.
For example if you have any changes for Boost.Format...

**Comment by EdSchouten** [comment]
That looks nifty. Thanks!

Unfortunately, the last couple of {days,weeks} I've been spending most of my time to get Node.js to work. Hacking on libuv, Google's V8, Node.js itself, etc., so I can't commit any time to sending in fixes right now. Sorry about that. :-(

If you happen to be interested in lending a hand here, that would be very much appreciated. Here's a link to the existing Boost package, still stuck at version 1.61:

https://github.com/NuxiNL/cloudabi-ports/tree/master/packages/boost

Getting these ported to more recent versions of Boost, cleaned up, upstreamed, etc. would be pretty sweet.

**Comment by jeking3** [comment]
Getting lexical_cast to work would be useful and also Boost.Test.

I question changing everything that wants to write to stdout, sending it to stderr however... that's not the intention of the author and I doubt anyone would accept a pull request with that in it, including myself.  There needs to be a better solution for stdout handling?

Those patches are nice insofar as they document pretty clearly some of the platform restrictions.  In Boost.Uuid I simply disabled tests what use std::cout or anything that uses lexical_cast, using those Boost.Predef macros.

**Comment by EdSchouten** [comment]
> There needs to be a better solution for stdout handling?

My observation is that the lack of `stdin` and `stdout` is something that you mainly notice when writing traditional UNIX command line tools. Tools that make use of `int main()`, use `getopt()` to parse their command line flags, iterate over file names arguments, open them, etc. Supporting that is outside the scope of CloudABI. With CloudABI, we try to build a programming environment in which people can write software that is fully dependency injected, using `cloudabi-run` and `program_main()`. Standard I/O streams don't fit inside this model; they shouldn't be exempt.

I think that this approach stimulates people to think about input and output of applications more carefully. Instead of trying to cram all of your tool's input and output over three streams, `cloudabi-run` gives you the freedom to create an arbitrary number of input and output streams. They can be optional, repeating, have names, etc.

You may notice that I've been somewhat inconsistent in this area. Why don't we support `stdin` and `stdout`, but do provide `stderr`? Why does `int main()` work in the first place, even though it's considered bad practice? Why not just throw a linker error that `program_main()` is missing? Well, that's exactly what the first versions of CloudABI did, but this turned out to be too strict.

- Almost all code out there depends on a logging sink, either directly or indirectly. Think of code that uses `assert()` or wants to print an error message prior to calling `abort()`. We do want to provide a facility to capture those messages somehow. This is why we provide `stderr`, although it's not hooked up to anything by default. CloudABI doesn't have any reserved file descriptors numbers (like 0, 1 and 2 on UNIX), so you need to initialise it explicitly (`FILE *x = fdopen(...); fswap(x, stderr); fclose(x);`).

- Build systems like GNU Autoconf and CMake essentially require that you support `int main()`.

Coming back to your specific example: the Boost.Uuid unit tests. I can imagine adding workarounds to the tests simply make them work on CloudABI makes little sense. That said, if a unit testing framework would support CloudABI properly, there wouldn't be any need to change unit tests in the first place, right? Some time ago I did manage to get `boost::tests` to work on CloudABI (see https://github.com/NuxiNL/cloudabi-ports/issues/29). Maybe it would be worth polishing that up?
